scalableminds/webknossos
# ND Datasets

This PR adds support for n-d datasets and annotations. Such datasets can be imported by using zarr.

See #7229 for restrictions and follow-up tasks regarding n-d support.

### URL of deployed dev instance (used for testing):
- https://d.webknossos.xyz

### Steps to test:
- go to https://d.webknossos.xyz/datasets/sample_organization/idr0101A-4d/view#1114,1075,17,0,0.977
  - was imported remotely with `https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0101A/13457227.zarr`
- use the t-slider in the toolbar to switch between different t values
- create a new annotation
- create skeleton nodes at different t values
  - edges should be half-transparent in that case
- create segments at different t values
- reload the page
- everything should be at the correct t values
- download/re-import skeletons via the skeleton tab
- download/re-upload whole annotation (not sure whether this is implemented yet?)

### Issues:
- fixes #4985

### Todo:

  - [x] accept additionalCoords as property in update actions
    - [x] skeletons. e.g., for createNode `{ "name": "createNode", "value": { "position": [...], "additionalCoords": [{ "name": "t", "value": 10 }], ...}}`
    - [x] volume. e.g., updateBucket and createSegment
      - [x] Save these in FossilDB. This requires changing Morton code calculation in https://github.com/scalableminds/webknossos-wrap ? Or only store xyz in morton code and additional coords afterward :heavy_check_mark: 
      	- Morton code is implemented currently. However, with only 64 Bits available, the coordinate bounds are very limited for high dimensions: Maximum coordinate = 2^(64/Dimensions) so with 5 dimensions already maximum coordinate value is 4096.
      	- [x] Validate that storing and getting them back actually works (did some weird regex parsing there)
      	- [ ] ~~Export/import for volume downloads/uploads~~
    - [x] allow to store additional coords in tracing (next to `editPosition`)
  - [ ] TracingStore: Move AdditionalCoordinateRequest.toProto to ProtoGeometryImplicits? Caused weird compiler errors at different places previously, maybe investigate
  - [x] Update DatasetArray
  - [x] Infer bounds of additional coordinates
  - [x] Make a clearer distinction between additionalCoordinate[request vs definition]. Maybe introduce new names?
  - [x] Import/export of nd-skeletons in NML (nodes should get properties, such as `additionalCoordinate-t="10"`)
  - [ ] ~~NML import or export of Volume annotations does not seem to work yet~~ -> change wkw
  - [x] consider writing tests (Added 13 tests :open_mouth: )
  - [x] On merging annotations / tracings etc., merge the additional coords
  - [x] Fix volume tracing bucket keys not needing migrations
  - [X] adapt data loading and rendering to higher dimensions
  - [X] adapt skeletons to higher dimensions
  - [x] test saving after backend was adapted
  - [x] encode additional coords in url
  - [x] respect in NML export
  - [x] tackle remaining todo comments
    - [x] fix rendering of skeletons
  - [x] fix cyclic dependencies

------
(Please delete unneeded items, merge only when none are left open)
